### PR TITLE
MNIST Dataset and ImageClassifier

### PR DESCRIPTION
**Pull request by yukw777**
All the commits included in this PR should be self-contained and easy to read one by one.

The first two commits fix some bugs I encountered along the way. The fix on `BaseSampler` may have unexpected consequences on other experiments, so let me know if there is another way to fix the issue.

Sample Experiment:
```yaml
!Experiment

name: mnist
resources:
  local:
    train_images: train-images-idx3-ubyte.gz
    train_labels: train-labels-idx1-ubyte.gz
    test_images: t10k-images-idx3-ubyte.gz
    test_labels: t10k-labels-idx1-ubyte.gz

pipeline:
  dataset: !MNISTDataset.from_path
    train_images_path: !@ train_images
    train_labels_path: !@ train_labels
    test_images_path: !@ test_images
    test_labels_path: !@ test_labels
  # or simply to download
  # dataset: !MNISTDataset.download

  model: !ImageClassifier
    encoder: !CNNEncoder
      input_channels: 1
      channels: [10, 20]
      kernel_size: [5, 5]
    linear: !MLPEncoder
      input_size: 8000  # 20 * 20 * 20
      output_size: 2048
      # output_activation: !torch.ReLU   <- due to https://github.com/asappresearch/flambe/issues/65, I could not set the activation layer
      n_layers: 2
      hidden_size: 4096
      # hidden_activation: !torch.ReLU   <- due to https://github.com/asappresearch/flambe/issues/65, I could not set the activation layer
    output_layer: !SoftmaxLayer
      input_size: 2048
      output_size: 10

  train: !Trainer
    dataset: !@ dataset
    train_sampler: !BaseSampler
      batch_size: 64
    val_sampler: !BaseSampler
    model: !@ model
    loss_fn: !torch.NLLLoss
    metric_fn: !Accuracy
    optimizer: !torch.Adam
      params: !@ train.model.trainable_params
    max_steps: 20
    iter_per_step: 50

```

### CODE REVIEW

File: flambe/vision/classification/model.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        The output layer, yields a probability distribution over targets
+    """
+    def __init__(self,
+                 encoder: Module,
+                 linear: Module,
+                 output_layer: Module) -> None:
+        super().__init__()
+
+        self.encoder = encoder
+        self.linear = linear
```
Comment (jeremyasapp) [inline flambe/vision/classification/model.py]: why the extra linear layer if there is already a linear output layer?

File: flambe/sampler/base.py [hunk 2]
```diff
@@ -177,7 +177,7 @@ def collate_fn(data: List[Tuple[torch.Tensor, ...]],
             sizes = [tensor.size() for tensor in tensors]
 
             if all(s == sizes[0] for s in sizes):
-                stacked_tensors = torch.stack(tensors).squeeze(1)
+                stacked_tensors = torch.stack(tensors)
```
Comment (jeremyasapp) [inline flambe/sampler/base.py]: Can you explain what this bug fixes? Is it for single channel images? To be honest, I'm not so sure if this squeeze is necessary.. I vaguely recall doing it for a reason, but I didn't add any comment, so hard to know now. Can you make sure the examples in the example folder run ok?

Comment (jeremyasapp) [reply flambe/sampler/base.py]: Oh it was for labels, which you seem to have addressed above

File: flambe/vision/classification/model.py (hunk 1 above)
Comment (yukw777) [reply flambe/vision/classification/model.py]: You could only have one fully connected layer with `SoftmaxLayer`, so i added this to add more layers. I could also modify `SoftmaxLayer` so that i could create multiple fully connected layers. Let me know which you prefer.

File: flambe/vision/classification/datasets.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        0x09: np.int8,
+        0x0b: np.dtype('>i2'),
+        0x0c: np.dtype('>i4'),
+        0x0d: np.dtype('>f4'),
+        0x0e: np.dtype('>f8'),
+    }
+
+    URL = "http://yann.lecun.com/exdb/mnist/"
+
+    def __init__(self,
```
[the remark this answers is not in the document]
Comment (yukw777) [reply flambe/vision/classification/datasets.py]: cool i will fix this.

File: flambe/vision/classification/model.py (hunk 1 above)
Comment (jeremyasapp) [reply flambe/vision/classification/model.py]: I guess to me if you wanted more fully connected layers, it seems like it would be better to have them in the encoder, but i understand that makes it weird with the CNNEncoder. I think it might make more sense to have the output layer be the one to add extra linear layers.

Comment (jeremyasapp) [reply flambe/vision/classification/model.py]: basically you could do:

```
output_layer: !MLPEncoder
    input_size: ...
    output_size: ...
    n_layers: 2
    output_activation: !torch.LogSoftmax
```
So i think the linear argument can be removed

Comment (jeremyasapp) [reply flambe/vision/classification/model.py]: Will fix in the next release. Let's merge as is
